server/app/assets/javascripts/session.ts

```diff
@@ -0,0 +1,395 @@
+   *
+   * @param data The value to check.
+   * @returns `true` if the value is a TimeoutData object, `false` otherwise.
+   */
+  private static isTimeoutData(data: unknown): data is TimeoutData {
+    // Check if the data is falsy or not an object. If so, it cannot be TimeoutData.
+    if (!data || typeof data !== 'object') {
+      return false
+    }
+
+    // Treat the data as a Record to allow accessing properties.
+    const d = data as Record<string, unknown>
+
+    // Check if all required properties exist and are of the correct type (number).
+    return (
+      typeof d.inactivityWarning === 'number' &&
+      typeof d.inactivityTimeout === 'number' &&
+      typeof d.totalWarning === 'number' &&
+      typeof d.totalTimeout === 'number' &&
+      typeof d.currentTime === 'number'
+    )
+  }
+
+  /**
+   * Retrieves the value of a cookie by its name.
+   *
+   * This function parses the `document.cookie` string to find the value
+   * associated with the given cookie name.
+   *
+   * @param name The name of the cookie to retrieve.
+   * @returns The string value of the cookie, or null if the cookie is not found.
+   */
+  private static getCookie(name: string): string | null {
+    // Prefix the document cookie string with a semicolon and a space to handle case when our cookie is the first.
+    const value = `; ${document.cookie}`
+
+    // Split the cookie string into an array of parts, using the cookie name as the delimiter.
+    // This assumes the cookie name is followed by an equals sign (=).
+    const parts = value.split(`; ${name}=`)
+
+    // Check if the cookie was found. If parts.length is 2, it means the cookie name
+    // was found, and the array contains a part before and a part after the cookie name.
+    if (parts.length === 2) {
+      // Extract the cookie value by taking the last element of the 'parts' array (the part after the cookie name),
+      // then splitting it by semicolons to separate the value from any subsequent cookie attributes
+      // (e.g., path, domain, etc.). Finally, take the first element (the actual value) and return it.
+      // If any of the split operations fail to return a value, return null.
+      return parts.pop()?.split(';').shift() || null
+    }
+
+    return null
+  }
+
+  /**
+   * Shows a warning modal of specified type.
+   * Only shows the modal if:
+   * - No other warning is currently shown
+   * - The modal element exists in the DOM
+   *
+   * Updates visibility tracking flags when showing a modal.
+   *
+   * @param type Type of warning to show (inactivity or total length)
+   */
+  private static showWarning(type: WarningType) {
+    // Check if any warning is already shown to prevent showing multiple dialogs
+    if (this.inactivityWarningShown || this.totalLengthWarningShown) return
+
+    // Get the appropriate modal element
+    const modalId =
+      type === WarningType.INACTIVITY
+        ? 'session-inactivity-warning-modal'
```

Comment:
suggestion (non-blocking): Since you use the strings `session-inactivity-warning` and `session-length-warning` multiple times you could store them in a const at the top of the file and note that they should be kept in sync with wherever they are set on the elements.

Reply:
done. Created an enum.